test/SILOptimizer/liveness_unit.sil

```diff
@@ -218,3 +223,191 @@ bb4(%phi : @owned $C):
+  %99 = tuple()
+  return %99 : $()
+}
+
+// One of the uses is a reborrowed inner borrow scope.
+//
+// CHECK-LABEL: testSSAInnerReborrowedPhi: ssa-liveness
+// CHECK: SSA lifetime analysis: %0 = argument of bb0 : $C
+// CHECK-NEXT: Incomplete liveness: Reborrowed inner scope
+// CHECK-NEXT: bb0: LiveWithin,
+// CHECK-NEXT: dead def: %0 = argument of bb0 : $C
+sil [ossa] @testSSAInnerReborrowedPhi : $@convention(thin) (@guaranteed C) -> () {
+bb0(%0 : @guaranteed $C):
+  test_specification "ssa-liveness @trace[0]"
+  debug_value [trace] %0 : $C
+  %borrow0 = begin_borrow %0 : $C
+  %aggregate = struct $PairC(%borrow0 : $C, %borrow0 : $C)
+  br bb1(%borrow0 : $C, %aggregate : $PairC)
+
+bb1(%reborrow : @guaranteed $C, %phi : @guaranteed $PairC):
+  %first = struct_extract %phi : $PairC, #PairC.first
+  %d = unchecked_ref_cast %first : $C to $D
+  %f = ref_element_addr %d : $D, #D.object  
+  %o = load [copy] %f : $*C
+  destroy_value %o : $C
+  end_borrow %reborrow : $C
+  %99 = tuple()
+  return %99 : $()
+}
+
+// Confusingly, deadPhi is the "last use", but it does not cause its
+// operands to be live-out of the predecessor block. This is because
+// liveness considers all phis to end liveness on the predecessor side
+// by default. Only the client can tell that a particular guaranteed phi
+// should actually extend liveness if it has no uses.
+//
+// CHECK-LABEL: testSSADeadGuaranteedPhi: ssa-liveness
+// CHECK: SSA lifetime analysis: %0 = argument of bb0 : $C
+// CHECK-NEXT: bb0: LiveOut,
+// CHECK-NEXT: bb2: LiveWithin,
+// CHECK-NEXT: bb1: LiveWithin,
+// CHECK-NEXT: regular user
+// CHECK: last user:
+// CHECK-SAME: br bb3
+// CHECK-NEXT: last user:   br bb3
+// CHECK-NEXT: end running
+sil [ossa] @testSSADeadGuaranteedPhi : $@convention(thin) (@guaranteed C) -> () {
+bb0(%0 : @guaranteed $C):
+  test_specification "ssa-liveness @trace[0]"
+  debug_value [trace] %0 : $C
+  cond_br undef, bb1, bb2
+
+bb1:
+  %d1 = unchecked_ref_cast %0 : $C to $D
+  br bb3(%d1 : $D)
+  
+bb2:
+  %d2 = unchecked_ref_cast %0 : $C to $D
+  br bb3(%d2 : $D)
+
+bb3(%phi : @guaranteed $D):
+  %99 = tuple()
+  return %99 : $()
+}
+
+// The phi is part of the guaranteed argument's simple live range.
+//
+// CHECK-LABEL: testSSADominatedPhi: ssa-liveness
+// CHECK: SSA lifetime analysis: %0 = argument of bb0 : $C
+// CHECK-NEXT: bb0: LiveOut,
+// CHECK-NEXT: bb2: LiveOut,
+// CHECK-NEXT: bb3: LiveWithin,
+// CHECK-NEXT: bb1: LiveOut,
+// CHECK-NEXT: regular user:   %{{.*}} = unchecked_ref_cast %0 : $C to $D
+// CHECK-NEXT: regular user:   br bb3
+// CHECK-NEXT: regular user:   %{{.*}} = load
+// CHECK-NEXT: regular user:   %{{.*}} = unchecked_ref_cast %0 : $C to $D
+// CHECK-NEXT: regular user:   br bb3
+// CHECK-NEXT: last user:   %{{.*}} = load
+// CHECK-NEXT: end running
+sil [ossa] @testSSADominatedPhi : $@convention(thin) (@guaranteed C) -> () {
+bb0(%0 : @guaranteed $C):
+  test_specification "ssa-liveness @trace[0]"
```

Comment:
Incidentally, this can also be written `@argument` and the `debug_value [trace]` instruction can be omitted.  This way is definitely more explicit.